File: lab2/utrushkin_daniil/src/fork_example.py
#!/usr/bin/env python3
import os
import time
import sys

def child_process(child_name):
    pid = os.getpid()
    ppid = os.getppid()
    print(f"{child_name}: PID={pid}, PPID={ppid}", flush=True)
    time.sleep(2)
    print(f"{child_name}: завершился", flush=True)
    return 0

def main():
    print(f"Родительский процесс: PID={os.getpid()}", flush=True)
    
    pid1 = os.fork()
    if pid1 == 0:
        sys.exit(child_process("child_A"))
    
    pid2 = os.fork()
    if pid2 == 0:
        sys.exit(child_process("child_B"))
    
    print(f"Родитель создал дочерние процессы: {pid1}, {pid2}", flush=True)
    
    for i in range(2):
        pid, status = os.wait()
        exit_code = os.WEXITSTATUS(status) if os.WIFEXITED(status) else -1
        print(f"Дочерний процесс {pid} завершился с кодом: {exit_code}", flush=True)
    
    print("Родительский процесс завершен", flush=True)

if __name__ == "__main__":
    main()

File: lab2/gr8sub2/СМОЛЬСКИЙ_ВЛАДИСЛАВ/src/fork_example.c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/types.h>


int main()
{
    pid_t pid1, pid2;

    printf("Parent started. PID: %d\n\n", getpid());
    fflush(stdout);

    pid1 = fork();
    if (pid1 < 0) // error
    {
        perror("fork");
        exit(1);
    }
    if (pid1 == 0) // child proc
    {
        printf("Child[0]: PID - %d, PPID - %d\n", getpid(), getppid());
        fflush(stdout);
        exit(0);
    }

    pid2 = fork();
    if (pid2 < 0) // error
    {
        perror("fork");
        exit(1);
    }
    if (pid2 == 0) // child proc
    {
        printf("Child[1]: PID - %d, PPID - %d\n", getpid(), getppid());
        fflush(stdout);
        exit(0);
    }

    int status;
    for (int i = 0; i < 2; ++i)
    {
        pid_t wpid = wait(&status);
        if (wpid < 0) // error
        {
            perror("wait");
            exit(1);
        }

        if (WIFEXITED(status))
        {
            printf("Parent: Chlild %d finished with code %d\n", wpid, WEXITSTATUS(status));
        }
        else
        {
            printf("Parent: Chlild %d finished abnormaly\n", wpid);
        }
        fflush(stdout);
    }

    printf("\nParent finished. PID: %d\n", getpid());

    return 0;
}